File: packages/cli/src/commands/mcp-setup.ts
/**
 * CLI command: matimo mcp setup
 *
 * Interactive config generator for MCP server.
 * Detects installed tools, scans for required auth env vars,
 * and outputs a ready-to-paste Claude Desktop / Cursor config.
 */

export async function mcpSetupCommand(): Promise<void> {
  console.info('\n🔨 Matimo MCP Setup\n');
  console.info('Scanning for installed tool packages...\n');

  try {
    // @ts-ignore - dynamic import of package resolved at runtime
    const { ToolLoader, extractAuthPlaceholders } = await import('@matimo/core');
    const loader = new ToolLoader();

    // Auto-discover installed packages
    const toolPaths = loader.autoDiscoverPackages();

    if (toolPaths.length === 0) {
      console.info('No @matimo/* tool packages found.');
      console.info('Install tools first: matimo install slack github\n');
      return;
    }

    // Load all tools
    const tools = loader.loadToolsFromMultiplePaths(toolPaths);

    console.info(`Found ${tools.size} tools across ${toolPaths.length} package(s):\n`);

    // Group tools by provider (extract from tool name prefix)
    const providers = new Map<string, string[]>();
    const authVars = new Set<string>();

    for (const [name, tool] of tools) {
      // Extract provider from tool name (e.g., "slack_send" → "slack")
      const provider = name.includes('_')
        ? name.split('_')[0]
        : name.includes('-')
          ? name.split('-')[0]
          : 'core';

      if (!providers.has(provider)) {
        providers.set(provider, []);
      }
      providers.get(provider)!.push(name);

      // Extract auth placeholders
      const placeholders = extractAuthPlaceholders(tool);
      for (const p of placeholders) {
        authVars.add(p);
      }
    }

    // Display providers and tools
    for (const [provider, toolNames] of providers) {
      console.info(`  📦 ${provider} (${toolNames.length} tools)`);
      for (const name of toolNames.slice(0, 5)) {
        console.info(`     • ${name}`);
      }
      if (toolNames.length > 5) {
        console.info(`     ... and ${toolNames.length - 5} more`);
      }
    }

    console.info('');

    // Display required env vars
    if (authVars.size > 0) {
      console.info('🔐 Required environment variables:\n');
      for (const v of authVars) {
        const value = process.env[v] || process.env[`MATIMO_${v}`];
        const status = value ? '✅' : '❌';
        console.info(`  ${status} ${v}`);
      }
      console.info('');
    }

    // Generate Claude Desktop config
    const envBlock: Record<string, string> = {};
    for (const v of authVars) {
      envBlock[v] = process.env[v] || '<your-token>';
    }

    const claudeConfig = {
      mcpServers: {
        matimo: {
          command: 'npx',
          args: ['matimo', 'mcp'],
          env: envBlock,
        },
      },
    };

    console.info('📋 Claude Desktop config (paste into Settings → Developer → MCP Servers):\n');
    console.info(JSON.stringify(claudeConfig, null, 2));

    // Generate Cursor config
    console.info('\n📋 Cursor config (paste into .cursor/mcp.json):\n');
    const cursorConfig = {
      mcpServers: {
        matimo: {
          command: 'npx',
          args: ['matimo', 'mcp'],
          env: envBlock,
        },
      },
    };
    console.info(JSON.stringify(cursorConfig, null, 2));

    // HTTP mode example
    console.info('\n📋 HTTP mode (for remote hosting / Docker):\n');
    const envExport = [...authVars].map((v) => `  ${v}=<your-token>`).join(' \\\n');
    console.info(`  ${envExport} \\`);
    console.info(`  MATIMO_MCP_TOKEN=<your-server-secret> \\`);
    console.info(`  npx matimo mcp --transport http --port 3000\n`);
  } catch (error) {
    console.error('❌ Setup failed:', error instanceof Error ? error.message : String(error));
    process.exit(1);
  }
}

File: python/packages/cli/src/matimo_cli/commands/mcp_setup.py
"""
``matimo mcp setup`` — generate MCP config for Claude Desktop / Cursor.

Mirrors: packages/cli/src/commands/mcp-setup.ts
"""
from __future__ import annotations

import json
import os
import re
import sys

_AUTH_PATTERNS = {"TOKEN", "SECRET", "KEY", "PASSWORD", "CREDENTIAL", "AUTH"}
_PLACEHOLDER_RE = re.compile(r"\{(\w+)\}")


def _is_auth_var(name: str) -> bool:
    upper = name.upper()
    return any(p in upper for p in _AUTH_PATTERNS)


def mcp_setup_command() -> None:
    print("\n🔨 Matimo MCP Setup\n")
    print("Scanning for installed tool packages…\n")

    try:
        from matimo.core.tool_loader import ToolLoader  # type: ignore[import-not-found]
    except ImportError:
        print("❌ matimo core not available. Install it first: pip install matimo", file=sys.stderr)
        sys.exit(1)

    try:
        loader = ToolLoader()
        tool_paths = loader.auto_discover_packages()

        if not tool_paths:
            print("No matimo-* tool packages found.")
            print("Install tools first: matimo install slack github\n")
            return

        tools = loader.load_tools_from_multiple_paths(tool_paths)
        print(f"Found {len(tools)} tools across {len(tool_paths)} package(s):\n")

        # Group by provider
        providers: dict[str, list[str]] = {}
        auth_vars: set[str] = set()

        for name, tool in tools.items():
            provider = name.split("_")[0] if "_" in name else name.split("-")[0] if "-" in name else "core"
            providers.setdefault(provider, []).append(name)

            # Extract auth placeholders
            exec_cfg = tool.execution
            raw = str(exec_cfg.model_dump()) if hasattr(exec_cfg, "model_dump") else str(exec_cfg)
            for m in _PLACEHOLDER_RE.finditer(raw):
                if _is_auth_var(m.group(1)):
                    auth_vars.add(m.group(1))

        for provider, tool_names in sorted(providers.items()):
            print(f"  📦 {provider} ({len(tool_names)} tools)")
            for tn in tool_names[:5]:
                print(f"     • {tn}")
            if len(tool_names) > 5:
                print(f"     … and {len(tool_names) - 5} more")

        print()

        # Display required env vars
        if auth_vars:
            print("🔐 Required environment variables:\n")
            for v in sorted(auth_vars):
                value = os.environ.get(v) or os.environ.get(f"MATIMO_{v}")
                status = "✅" if value else "❌"
                print(f"  {status} {v}")
            print()

        # Generate configs
        env_block = {v: os.environ.get(v) or os.environ.get(f"MATIMO_{v}") or "<your-token>" for v in sorted(auth_vars)}

        claude_config = {
            "mcpServers": {
                "matimo": {
                    "command": "matimo",
                    "args": ["mcp"],
                    "env": env_block,
                }
            }
        }

        print("📋 Claude Desktop config (paste into Settings → Developer → MCP Servers):\n")
        print(json.dumps(claude_config, indent=2))

        print("\n📋 Cursor config (paste into .cursor/mcp.json):\n")
        print(json.dumps(claude_config, indent=2))

        print("\n📋 HTTP mode (for remote hosting / Docker):\n")
        env_lines = "\n".join(f"  {v}=<your-token>" for v in sorted(auth_vars))
        print(f"{env_lines}")
        print("  MATIMO_MCP_TOKEN=<your-server-secret>")
        print("  matimo mcp --transport http --port 3000\n")

    except Exception as exc:
        print(f"❌ Setup failed: {exc}", file=sys.stderr)
        sys.exit(1)
